candid_rust/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+                    Label::Id(i) => format!("id_{}", i),
+                    Label::Named(n) => n.clone(),
+                    Label::Unnamed(i) => format!("unnamed_{}", i),
+                };
+                let label = syn::Ident::new(&label_str, span.span());
+                let ty = idl_type_usage_to_rust(context, &f.typ, &span.span(), true)?;
+
+                all_fields.push(quote! {
+                    pub #label : #ty
+                });
+            }
+
+            quote! {
+                struct #id {
+                    #(#all_fields),*
+                }
+            }
+        }
+        ty => {
+            let usage = idl_type_usage_to_rust(context, ty, span, false)?;
+            quote! { type #id = #usage ; }
+        }
+    });
+}
+
+fn idl_type_usage_to_rust(
+    context: &Context,
+    ty: &IDLType,
+    span: &dyn Spanned,
+    should_box: bool,
+) -> Result<TokenStream> {
+    Ok(match ty {
+        IDLType::PrimT(prim_t) => match prim_t {
+            // TODO: support bigint.
+            PrimType::Nat => quote! { u128 },
+            PrimType::Nat8 => quote! { u8 },
+            PrimType::Nat16 => quote! { u16 },
+            PrimType::Nat32 => quote! { u32 },
+            PrimType::Nat64 => quote! { u64 },
+            PrimType::Int => quote! { i128 },
+            PrimType::Int8 => quote! { i8 },
+            PrimType::Int16 => quote! { i16 },
+            PrimType::Int32 => quote! { i32 },
+            PrimType::Int64 => quote! { i64 },
+            PrimType::Float32 => quote! { f32 },
+            PrimType::Float64 => quote! { f64 },
+            PrimType::Bool => quote! { bool },
+            PrimType::Text => quote! { String },
+            PrimType::Null => quote! { () },
+            PrimType::Empty => quote! { () },
+
+            PrimType::Reserved => {
+                return Err(errors::Error::Custom(
+                    "Unsupported PrimType: Reserved".to_string(),
+                ));
+            }
+        },
+
+        IDLType::OptT(sub_t) => {
+            let subquote = idl_type_usage_to_rust(context, sub_t, span, false)?;
+            quote! { Option<#subquote> }
+        }
+
+        IDLType::VecT(sub_t) => {
+            let subquote = idl_type_usage_to_rust(context, sub_t, span, false)?;
+            quote! { Vec<#subquote> }
+        }
+
+        IDLType::VarT(name) => {
+            let id = syn::Ident::new(name, span.span());
+            if should_box && context.is_record(name) {
+                quote! { Box<#id> }
+            } else {
+                quote! { #id }
+            }
+        }
+
+        x => return Err(errors::Error::Custom(format!("Unsupported type: {:?}", x))),
+    })
+}
+
+pub fn idl_prog_into_stream(prog: IDLProg) -> Result<TokenStream> {
+    let mut context = Context::new(&prog);
+
+    let mut all_declarations = Vec::<TokenStream>::new();
+    for dec in &prog.decs {
+        context.add_declaration(dec);
+    }
+
+    for dec in &prog.decs {
```

Comment:
Starting directly from `dec` may include unused type defs in the generated code. Now that we have context, a more robust way is to start from actor, and gather the mentioned type definitions into context.

Reply:
I think that's fine for now. This would be a larger change right now than required. I can do it as an optimization later.